ci: add github action workflows

laojianzi opened the pull request:
Trying to solve: #2582 

Merge before: 

- open [Setting > Secrets](https://github.com/gin-gonic/gin/settings/secrets/actions) and set you'r gitter room id to `GITTER_ROOM_ID `
- Merge #2595

**If the master branch changes, or if a problem occurs that does not necessarily occur, you can choose to re-run the github action job.**

Differences from Travis CI:

- Multiple os `ubuntu-latest`, `macos-latest`, `windows-latest`
- No go master version
- Sending gitter notifications using curl

---

### Some of the problems I've encountered since using GitHub Action

- `TestUnixSocket` windows cannot read `/tmp/unix_unit_test` **(fixed: #2595)**
- `TestPathCleanMallocs` have a certain chance of failure **(I don't know how to repair it :confused: )**
>     path_test.go:85: 
>        	Error Trace:	path_test.go:85
>        	Error:      	Not equal: 
>        	            	expected: float64(2010)
>        	            	actual  : int(0)
>        	Test:       	TestPathCleanMallocs
- `TestUnixSocket` have a certain chance of failure **(I don't know how to repair it :confused: )**
> Get http://127.0.0.1:49379/test: dial tcp 127.0.0.1:49379: bad file descriptor

thinkerou commented:
`open Setting > Secrets and set you'r gitter room id to GITTER_ROOM_ID`
-- now only @javierprovecho can set it

laojianzi commented:
> `open Setting > Secrets and set you'r gitter room id to GITTER_ROOM_ID`
> -- now only @javierprovecho can set it

https://github.com/gin-gonic/gin/blob/b01605bb5b43dbf33781970af5ad6633e5549fd1/.travis.yml#L47

@thinkerou Use `7f95bf605c4d356372f4`, as in `.travis.yml`.

What do you think?

thinkerou commented:
> > `open Setting > Secrets and set you'r gitter room id to GITTER_ROOM_ID`
> > -- now only @javierprovecho can set it
> 
> https://github.com/gin-gonic/gin/blob/b01605bb5b43dbf33781970af5ad6633e5549fd1/.travis.yml#L47
> 
> @thinkerou Replace `GITTER_ROOM_ID` with `7f95bf605c4d356372f4`, as in `.travis.yml`.
> 
> What do you think?

Yes, OK! thanks.

thinkerou commented on .github/workflows/gin.yml:
```diff
@@ -0,0 +1,63 @@
+        run: make test
+
+      - name: Upload coverage to Codecov
+        if: success() && matrix.go == 1.15 && matrix.os == 'ubuntu-latest' && matrix.test-tags == 'nomsgpack'
```
why `matrix.go == 1.15` here?

laojianzi replied:
If the `go test` is successful, the coverage is theoretically the same every time

So here we just need to avoid multiple uploads and choose one of them

thinkerou replied:
@laojianzi ok, thank your reply.

laojianzi replied:
1.12.x: https://travis-ci.org/github/gin-gonic/gin/jobs/753866331#L1150
1.13.x: https://travis-ci.org/github/gin-gonic/gin/jobs/753866332#L1150
...

@thinkerou In Travis CI though, it seems that the go test coverage is uploaded for every go version, so I can update it if you want.

thinkerou replied:
@laojianzi yes, please update, and remove `.travis.yml` and update gitter-room-id,thanks

laojianzi replied:
@thinkerou updated

laojianzi commented:
https://github.com/gin-gonic/gin/runs/1706106769?check_suite_focus=true#step:6:32
```
--- FAIL: TestLoadHTMLFilesDebugMode (0.00s)
make: *** [test] Error 1
    debug_test.go:144: 
        	Error Trace:	debug_test.go:144
        	            				asm_amd64.s:1357
        	Error:      	Received unexpected error:
        	            	read |0: bad file descriptor
        	Test:       	TestLoadHTMLFilesDebugMode
```

`bad file descriptor` often occurs, suspected to be related to the use of `goroutine` in testing (speculation without any evidence)

thinkerou reviewed (dismissed):
lgtm, thanks!

thinkerou commented:
@appleboy please review the pr, thanks!

`continuous-integration/travis-ci Expected — Waiting for status to be reported`
why still required?

laojianzi commented:
@thinkerou Maybe it has something to do with this

![image](https://user-images.githubusercontent.com/42930263/104690394-5412b380-573f-11eb-95f9-d000b3565ff6.png)

laojianzi commented:
https://github.com/gin-gonic/gin/blob/540b1eff7069128df5d95e09968569f2266d9a6a/.github/PULL_REQUEST_TEMPLATE.md#L4
https://github.com/gin-gonic/gin/blob/540b1eff7069128df5d95e09968569f2266d9a6a/CONTRIBUTING.md#L11
https://github.com/gin-gonic/gin/blob/46ddd4259cac975be1eb11b4f1192264f582db16/README.md#L5

@thinkerou I think these need to be replaced with GitHub Action, do they need to be updated here?

thinkerou commented:
> https://github.com/gin-gonic/gin/blob/540b1eff7069128df5d95e09968569f2266d9a6a/.github/PULL_REQUEST_TEMPLATE.md#L4
> 
> 
> https://github.com/gin-gonic/gin/blob/540b1eff7069128df5d95e09968569f2266d9a6a/CONTRIBUTING.md#L11
> 
> 
> https://github.com/gin-gonic/gin/blob/46ddd4259cac975be1eb11b4f1192264f582db16/README.md#L5
> 
> @thinkerou I think these need to be replaced with GitHub Action, do they need to be updated here?

yes, please update, thanks!

laojianzi commented:
@thinkerou @appleboy okay, plz review, thanks!

thinkerou commented:
> @thinkerou Maybe it has something to do with this, only @javierprovecho can set it?
> 
> ![image](https://user-images.githubusercontent.com/42930263/104690394-5412b380-573f-11eb-95f9-d000b3565ff6.png)

yes, please @javierprovecho review and set, thanks!

appleboy commented on .github/workflows/gin.yml:
```diff
@@ -0,0 +1,62 @@
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+        go: [1.12, 1.13, 1.14, 1.15]
```
also add `1.16`

appleboy commented:
@laojianzi Please also fix the conflicts?

thinkerou commented:
> @laojianzi Please also fix the conflicts?

we should remove `.travis.yml`?

laojianzi commented:
Thank

@appleboy add go version 1.16 to ci and fixed the conflicts
@thinkerou I didn't think it was necessary to keep .travis.yml after using github actions, so I deleted it

Please review again

thinkerou commented:
> Thank
> 
> @appleboy add go version 1.16 to ci and fixed the conflicts
> @thinkerou I didn't think it was necessary to keep .travis.yml after using github actions, so I deleted it
> 
> Please review again

ok, thanks!
wait @javierprovecho solve this: `continuous-integration/travis-ci Expected — Waiting for status to be reported`

appleboy commented:
@thinkerou Move the PR to v1.8 milestone?

thinkerou commented:
> @thinkerou Move the PR to v1.8 milestone?

OK, merge to marster before removing `continuous-integration/travis-ci Expected — Waiting for status to be reported`

thinkerou commented:
@laojianzi please fix conflict and we will merge it, thanks!

laojianzi commented:
@thinkerou @appleboy https://golang.org/pkg/net/#pkg-note-BUG

Should tests requiring fd be skipped when run in windows?

appleboy commented:
@laojianzi We can skip the windows testing first and maybe create another PR to track the testing on Windows.

appleboy commented on .github/workflows/gin.yml:
```diff
@@ -0,0 +1,62 @@
+  test:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
```
remove windows-latest

appleboy commented:
Waiting for @thinkerou approval.

thinkerou reviewed (approved):
lgtm, thanks a lot!
